### PR TITLE
Support PHP 8.0

### PR DESCRIPTION
**Pull request by deguif**
#### Done in this PR:

- [x] Update composer constraint to allow PHP 8.0.
- [x] Update travis configuration in order to run tests on PHP 8.0 (nightly).
- [x] Update changelog

#### Done in previous PRs:
- [x] Deprecate `Elastica\Query\Match` class and add a new `Elastica\Query\MatchQuery` class as match is now a reserved keyword. [#1799]
- [x] Update phpunit version in order to be compatible with PHP 8.0. [#1811]
- [x] Bump `aws/aws-sdk-php` package version to fix a bug with PHP 8.0. [#1798]

#### Done in external PRs:
- [x] Wait for `aws/aws-sdk-php` to be compatible with PHP 8. [https://github.com/aws/aws-sdk-php/pull/2079]
- [ ] Wait for `elasticsearch/elasticsearch` to be compatible with PHP 8.0 and remove `--ignore-platform-reqs` composer option.

### CODE REVIEW

File: src/Util.php [hunk 1]
```diff
@@ -221,7 +221,7 @@ public static function getParamName($class)
 
         $parts = \explode('\\', $class);
         $last = \array_pop($parts);
-        $last = \preg_replace('/Query$/', '', $last); // for BoolQuery
+        $last = \preg_replace('/(Query|_)$/', '', $last); // for BoolQuery and Match_
```
Comment (deguif) [inline src/Util.php:224]: Replaced `_` suffix here, but the method can be overriden to return `match` instead. WDYT @ruflin ?

File: src/QueryBuilder/DSL/Query.php [hunk 2]
```diff
@@ -72,9 +72,9 @@ public function getType(): string
      *
      * @param mixed $values
      */
-    public function match(?string $field = null, $values = null): Match
+    public function match(?string $field = null, $values = null): Match_
```
Comment (deguif) [inline src/QueryBuilder/DSL/Query.php:75]: BC break here, but impact should be very low.

**Comment by deguif** [comment]
One failure is remaining on PHP 8.0 but it's related to aws SDK.
See https://github.com/aws/aws-sdk-php/pull/2079 for details and the fix that should be merged soon (I hope).

File: src/Query/Match.php [hunk 3]
```diff
@@ -2,180 +2,17 @@
 
 namespace Elastica\Query;
 
+\trigger_error('Elastica\Query\Match is deprecated. Use Match instead. From PHP 8 match is reserved word and this class will be removed in further Elastica releases', \E_USER_DEPRECATED);
```
Comment (jdreesen) [inline src/Query/Match.php:5]:
```suggestion
\trigger_error('Elastica\Query\Match is deprecated. Use Match_ instead. From PHP 8 match is reserved word and this class will be removed in further Elastica releases', \E_USER_DEPRECATED);
```

Comment (deguif) [reply src/Query/Match.php]: Thanks @jdreesen

**Comment by ruflin** [comment]
@deguif Thanks for getting this started. We had in the past a similar issue with a naming conflict with the `Bool` query. What we did back then is rename it to `BoolQuery` instead: https://github.com/ruflin/Elastica/blob/master/src/Query/BoolQuery.php I wonder if we should follow the same pattern here instead of using the `_` for consistency? `MatchQuery`?

File: .travis.yml [hunk 4]
```diff
@@ -15,20 +15,25 @@ matrix:
...
-      env: RUN_TEST=true  ES_VERSION=73
+      env: RUN_TEST=true ES_VERSION=73
     - php: 7.4
-      env: RUN_TEST=true  ES_VERSION=73
+      env: RUN_TEST=true ES_VERSION=73
+    - php: nightly
+      env: RUN_TEST=true ES_VERSION=73 COMPOSER_FLAGS="--ignore-platform-reqs"
     # Testing the library with different ES versions
-    - php: 7.2
+    - php: 7.3
```
Comment (ruflin) [inline .travis.yml:24]: Any chance you could push the changes for PHP 7.3 into a separate PR sow e can get it in quickly? Then this PR gets smaller and we know potential failures are only related to these changes.

File: src/Query/Match.php [hunk 5]
```diff
@@ -2,180 +2,17 @@
...
 namespace Elastica\Query;
 
+\trigger_error('Elastica\Query\Match is deprecated. Use Match_ instead. From PHP 8 match is reserved word and this class will be removed in further Elastica releases', \E_USER_DEPRECATED);
+
 /**
  * Match query.
  *
- * @author F21
- * @author WONG Wing Lun <luiges90@gmail.com>
+ * This class is for backward compatibility reason. For PHP 8 and above use Match_ as Match is reserved.
```
Comment (ruflin) [inline src/Query/Match.php:10]: Nice, so we can keep the old file around for compatiblity but mark it deprecated. This is great.

File: src/Util.php (hunk 1 above)
Comment (ruflin) [reply src/Util.php:224]: Not sure I follow? If we go with `Query` suffix instead of `_`, will it just work?

File: tests/PipelineTest.php [hunk 6]
```diff
@@ -66,7 +66,8 @@ public function testPipelineCreate(): void
 
         $result = $pipeline->create();
 
-        $this->assertContains('acknowledged', $result->getData());
+        $this->assertArrayHasKey('acknowledged', $result->getData());
```
Comment (ruflin) [inline tests/PipelineTest.php:69]: Are these changes related to PHP 8? The new phpunit version? This this new version also work in 7.x? 

This feeds into the same question like the php 7.3 part above: Could we have these change in a separate PR to make it clear they are not related to the PHP 8.0 changes?

File: src/Util.php (hunk 1 above)
Comment (deguif) [reply src/Util.php:224]: Yes

**Comment by ruflin** [comment]
@deguif This will also need an update of the CHANGELOG and I wonder if we should also update the readme that it now running at least tests with PHP 8.0?

File: tests/PipelineTest.php (hunk 6 above)
Comment (deguif) [reply tests/PipelineTest.php:69]: yes, so increase of php version should be merged before increasing phpunit version to 9.3 which supports php 8.0 but doesn't support php 7.2 anymore.

Comment (ruflin) [reply tests/PipelineTest.php:69]: It seems like it is also compatible with 7.3, so if we jump to 7.3, we can go to 9 first in a separate PR and do this changes?

File: src/QueryBuilder/DSL/Query.php (hunk 2 above)
Comment (BackEndTea) [reply src/QueryBuilder/DSL/Query.php:75]: Not for users running php 7.4 and up.

As for them if they extend it they can speciyf their method returns the `Match` class.

Comment (deguif) [reply src/QueryBuilder/DSL/Query.php:75]: It won't impact users on 7.4 just for covariance.
But probability that users care about this return is very low ;)
By care I mean `instanceof` or anything that is tied to the previous class name

**Comment by GrahamCampbell** [comment]
> Bump aws/aws-sdk-php package version to fix a bug with PHP 8.0 (aws/aws-sdk-php#2079).

**Comment by GrahamCampbell** [comment]
> Bump aws/aws-sdk-php package version to fix a bug with PHP 8.0 (aws/aws-sdk-php#2079).

There are definitely more similar bugs like the one that I fixed. Amazon's job is to fix that, though. I have just fixed the one that affected me.

**Comment by GrahamCampbell** [comment]
Also, Guzzle, a dependency of the AWS SDK is not ready for PHP 8 yet.

**Comment by ruflin** [comment]
Is there option to exclude the AWS part for 8.0 for now?

**Comment by GrahamCampbell** [comment]
Guzzle is now ready, btw. 7.2.0 supports PHP 8.0.

File: tests/ClientConfigurationTest.php [hunk 7]
```diff
@@ -15,9 +15,9 @@ class ClientConfigurationTest extends TestCase
     public function testInvalidDsn(): void
     {
         $this->expectException(\Elastica\Exception\InvalidException::class);
-        $this->expectExceptionMessage("DSN 'test:0' is invalid.");
+        $this->expectExceptionMessage("DSN ':0' is invalid.");
 
-        ClientConfiguration::fromDsn('test:0');
```
Comment (deguif) [inline tests/ClientConfigurationTest.php:20]: Since PHP 8.0 RC1 a 0 port number is considered valid: See https://github.com/php/php-src/pull/6152
For details see how it behaves on other PHP versions: https://3v4l.org/ZYhJc

**Comment by deguif** [comment]
@ruflin I think this PR is now ready.
Testing on PHP 8.0 is now green with all the changes done in previous PRs.

I will later update the PHP version in travis configuration to 8.0 and also remove the composer flag `--ignore-platform-reqs` when `elasticsearch/elasticsearch` will officially support PHP 8.0.

**Comment by ruflin** [comment]
@deguif Nice! Thank you!
